docs/differences_from_aws.md

```diff
@@ -0,0 +1,101 @@
+# Differences from AWS Architecture
+
+This document explains how GCP Ingestion differs from the AWS implementation.
+
+[AWS Data Platform Architecture](https://mana.mozilla.org/wiki/display/SVCOPS/Telemetry+-+Data+Pipeline+Architecture)
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+
+- [Replace Heka Message Format with PubSub Message Format](#replace-heka-message-format-with-pubsub-message-format)
+- [Replace EC2 Edge with Kubernetes Edge](#replace-ec2-edge-with-kubernetes-edge)
+- [Replace Kafka with PubSub](#replace-kafka-with-pubsub)
+- [Replace Hindsight Data Warehouse Loaders with Dataflow](#replace-hindsight-data-warehouse-loaders-with-dataflow)
+- [Replace S3 with Cloud Storage](#replace-s3-with-cloud-storage)
+- [Messages Always Delivered to Message Queue](#messages-always-delivered-to-message-queue)
+- [Landfill is Downstream from Message Queue](#landfill-is-downstream-from-message-queue)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+
+## Replace Heka Message Format with PubSub Message Format
+
+Heka messages are protobuf, and have a special framing format. PubSub messages
+are newline delimited JSON. The PubSub message format is being used both because
+it works well with PubSub and also because it can be read using Dataflow the
+same way from PubSub and Cloud Storage with minimal effort.
+
+## Replace EC2 Edge with Kubernetes Edge
+
+The AWS data platform uses EC2 instances running an NGinX module to encode HTTP
+requests as Heka messages, writes them to Kafka using librdkafka and directly to
+files on disk. librdkafka handles buffering and batching when writing to Kafka.
+Files on disk are rotated with cron and uploaded to S3. On shutdown files are
+forcefully rotated and uploaded. The sizing of the EC2 instance cluster is
+effectively static, but is configured to scale up if needed.
+
+The EC2 instances have been replaced with a Kubernetes cluster. This decision
+was made by the Cloud Operations team to simplify operational support for them.
+
+The NGinX module has been replaced by an HTTP service running in Docker. A
+number of factors informed the decision to rewrite the edge, including:
+
+## Replace Kafka with PubSub
+
+Comparison:
+
+|                | Kafka in AWS                    | PubSub                      |
+|----------------|---------------------------------|-----------------------------|
+| Managed by     | Ops                             | Google                      |
+| Access control | Security groups, all-or-nothing | Cloud IAM, per-topic        |
+| Scaling        | Manual                          | Automatic                   |
+| Cost           | Per EC2 instance                | Per GB, min charge 1 KB/req |
+| Data Storage   | 2-3 days, depends on throughput | 7 days for unacknowledged   |
```

Comment:
2-3 days here is again something we configured, not a limitation of Kafka as a technology. 

Perhaps the important difference is that storage is fixed and exhaustible for Kafka, but auto-scaling for PubSub.

Reply:
I changed this to "Configured in GB per EC2 instance" because I feel like that cuts to the heart of the issue. PubSub's advantage is that it's an amount of time, not storage.